From 7b0176ccd05ef34a98a1729411eadbcf50cf1cb8 Mon Sep 17 00:00:00 2001
From: Eric Sauvageau <merlin@asuswrt-merlin.net>
Date: Sun, 23 Aug 2026 20:41:22 -0400
Subject: [PATCH] httpd: More wireless log fixes for Repeater and Media Bridge
 modes

- Fixed reporting while in Media Bridge and Repeater modes
- Improved reporting while RP/MB is disconnected or in an error state

(cherry picked from commit da49a89a598dc7704eac4019c7f4f788b93f2c86)
---
 .../router/httpd/sysdeps/web-broadcom-am.c    | 51 ++++++++++++-------
 1 file changed, 32 insertions(+), 19 deletions(-)

diff --git a/release/src/router/httpd/sysdeps/web-broadcom-am.c b/release/src/router/httpd/sysdeps/web-broadcom-am.c
index 2130f7ff18..f6b2d8034a 100644
--- a/release/src/router/httpd/sysdeps/web-broadcom-am.c
+++ b/release/src/router/httpd/sysdeps/web-broadcom-am.c
@@ -294,14 +294,29 @@ wl_status_array(int eid, webs_t wp, int argc, char_t **argv, int unit)
 	char tmp[128], prefix[] = "wlXXXXXXXXXX_";
 	char *name;
 
-	snprintf(prefix, sizeof(prefix), "wl%d.1_", unit);
+	if (repeater_mode()
+#ifdef RTCONFIG_PROXYSTA
+	    || is_psr(unit)
+	    || dpsr_mode()
+	    || is_psta(unit)
+	    || rp_mode()
+#ifdef RTCONFIG_DPSTA
+	    || dpsta_mode()
+#endif
+#endif
+	) {
+		snprintf(prefix, sizeof(prefix), "wl%d_", unit);
+	} else {
+		snprintf(prefix, sizeof(prefix), "wl%d.1_", unit);
+	}
+
 	name = nvram_safe_get(strcat_r(prefix, "ifname", tmp));
 
 	if ((ret = wl_ioctl(name, WLC_GET_BSSID, &bssid, ETHER_ADDR_LEN)) == 0) {
 		/* The adapter is associated. */
 		*(uint32*)buf = htod32(WLC_IOCTL_MAXLEN);
 		if ((ret = wl_ioctl(name, WLC_GET_BSS_INFO, buf, WLC_IOCTL_MAXLEN)) < 0) {
-			retval += websWrite(wp, "\"?\",\"?\",\"?\",\"?\",\"?\",\"?\",");
+			retval += websWrite(wp, "\"?\",\"0\",\"0\",\"0\",\"0\",\"?\",\"-1\",");
 			return retval;
 		}
 #ifdef RTCONFIG_HND_ROUTER_AX
@@ -314,12 +329,12 @@ wl_status_array(int eid, webs_t wp, int argc, char_t **argv, int unit)
 		    dtoh32(bi->version) == LEGACY_WL_BSS_INFO_VERSION)
 			retval += dump_bss_info_array(eid, wp, argc, argv, bi, name);
 		else
-			retval += websWrite(wp, "\"<error>\",\"\",\"\",\"\",\"\",\"\",");
+			retval += websWrite(wp, "\"<error>\",\"0\",\"0\",\"0\",\"0\",\"\",\"-1\",");
 	} else {
 		retval += websWrite(wp, "\"Not associated. Last with ");
 
 		if ((ret = wl_ioctl(name, WLC_GET_SSID, &ssid, sizeof(wlc_ssid_t))) < 0) {
-			retval += websWrite(wp, "<unknown>\",\"\",\"\",\"\",\"\",\"\",");
+			retval += websWrite(wp, "<unknown>\",\"0\",\"0\",\"0\",\"0\",\"\",\"-1\",");
 			return 0;
 		}
 
@@ -328,7 +343,7 @@ wl_status_array(int eid, webs_t wp, int argc, char_t **argv, int unit)
 		if (str_escape_quotes(ssidbuf, ssidbuftmp, sizeof(ssidbuf)) == 0 )
 			strlcpy(ssidbuf, ssidbuftmp, sizeof(ssidbuf));
 
-		retval += websWrite(wp, "%s\",\"\",\"\",\"\",\"\",\"\",", ssidbuf);
+		retval += websWrite(wp, "%s\",\"0\",\"0\",\"0\",\"0\",\"\",\"-1\",", ssidbuf);
 	}
 
 	return retval;
@@ -397,7 +412,7 @@ ej_wl_unit_status_array(int eid, webs_t wp, int argc, char_t **argv, int unit)
 	ret += websWrite(wp, "dataarray%d = [", unit);
 
 	if (nvram_match(strcat_r(prefix, "mode", tmp), "wds"))
-		ret += websWrite(wp, "\"\",\"\",\"\",\"\",\"%d\",\"\",", wl_control_channel(unit));
+		ret += websWrite(wp, "\"\",\"0\",\"0\",\"0\",\"%d\",\"\",", wl_control_channel(unit));
 	else
 		ret += wl_status_array(eid, wp, argc, argv, unit);
 
@@ -419,24 +434,22 @@ ej_wl_unit_status_array(int eid, webs_t wp, int argc, char_t **argv, int unit)
 		ret += websWrite(wp, "\"Stations\"");
 		noclients = 1;
 	}
-	else if (nvram_match(strcat_r(prefix, "mode", tmp), "wet"))
-	{
-#ifdef RTCONFIG_WIRELESSREPEATER
-		if ((nvram_get_int("sw_mode") == SW_MODE_REPEATER)
-			&& (nvram_get_int("wlc_band") == unit))
-			sprintf(prefix, "wl%d.%d_", unit, 1);
-#endif
-		ret += websWrite(wp, "\"Repeater ( SSID local: %s )\"", nvram_safe_get(strcat_r(prefix, "ssid", tmp)));
-	}
 #ifdef RTCONFIG_PROXYSTA
-	else if (nvram_match(strcat_r(prefix, "mode", tmp), "psta"))
+	else if ((sw_mode() == SW_MODE_AP) &&
+		     (nvram_get_int("wlc_psta") == 1) &&
+		     (nvram_get_int("wlc_band") == unit))
 	{
-		if ((nvram_get_int("sw_mode") == SW_MODE_AP) &&
-			(nvram_get_int("wlc_psta") == 1) &&
-			(nvram_get_int("wlc_band") == unit))
 		ret += websWrite(wp, "\"Media Bridge\"");
 	}
 #endif
+	else if (nvram_match(strcat_r(prefix, "mode", tmp), "wet"))
+	{
+		if (nvram_get_int("wlc_band") == unit)
+			sprintf(prefix, "wl%d.%d_", unit, 1);
+
+		ret += websWrite(wp, "\"Repeater ( SSID local: %s )\"", nvram_safe_get(strlcat_r(prefix, "ssid", tmp, sizeof(tmp))));
+	}
+
 
 // Close dataarray
 	ret += websWrite(wp, "];\n");
-- 
2.25.1

